src/optimizers/quantizer.cpp

```diff
@@ -0,0 +1,187 @@
+           * pow(base,  // revert from log space to normal FP represtation
+                 clip(floor(log(abs(_1 / S) * mult) / log(base)),  // get the quantization center
+                      (float) numCenters)),                                // clip
+      res);
+}
+
+/* Quantize all the parameters (except bias, unless enabled via --quantize-bias)
+ * Quantization only works if we store the quantization error residual
+ * The stored residual will be added for the next quantization
+ * @param graph is the model graph to be quantized (in-place) 
+ */
+void ModelQuantizer::quantize(Ptr<ExpressionGraph> graph) {
+  // reserve tensor for error feedback mechanism
+  if(!errorResidual_) {
+    LOG(info, " EXPERIMENTAL: quantize the model to {}-bits", bits_);
+
+    int numElements = (int)graph->params()->vals()->size();
+    auto allocator = New<TensorAllocator>(graph->getBackend());
+    allocator->reserveExact(graph->params()->vals()->memory()->size());
+    allocator->allocate(errorResidual_, {1, numElements});
+
+    allocators_.push_back(allocator);
+    firstError_ = true;
+  }
+
+  {
+    // apply error feedback mechanism
+    using namespace functional;
+    // add the previous error residual to the current model
+    Element(_1 += _2, graph->params()->vals(), errorResidual_);
+    errorResidual_->copyFrom(graph->params()->vals());
+  }
+
+  for(auto p : *graph->params()) {
+    // optionally quantize bias
+    if(quantBias_ || p->val()->shape()[0] > 1)
+      quantizeImpl(p->val());
+  }
+
+  // get new error residual. Skip the first one.
+  if (!firstError_) {
+    using namespace functional;
+    Element(_1 -= _2, errorResidual_, graph->params()->vals());
+  }
+  else {
+    errorResidual_->set(0);
+    firstError_ = false;
+  }
+}
+
+
+/* Tensor quantization implementation.
+ * @param t is the tensor to be quantized (in-place)
+ */
+void ModelQuantizer::quantizeImpl(Tensor t) {
+  if(!tempVar_) {
+    // init the swap tensor
+    auto allocator = New<TensorAllocator>(t->getBackend());
+    allocator->reserveExact(sizeof(float));
+    allocator->allocate(tempVar_, {1, 1});
+    allocators_.push_back(allocator);
+  }
+
+  // init additional tensor for scaling optimization
+  if(!delta_ && optSteps_ > 0) {
+    int msize = (int) t->size();
+    auto allocator = New<TensorAllocator>(t->getBackend());
+    allocator->reserveExact(msize * sizeof(float));
+    allocator->allocate(delta_, {1, msize});
+    allocators_.push_back(allocator);
+  }
+  
+  Tensor q = delta_->subtensor(0, t->size());  // to store the quantized t
+  Tensor tflat = t->subtensor(0, t->size());   // flatten t for reduce
+
+  // scaling factor S
+  float S = 0.0f;
+  // get intial scaling factor (S) based on max element in Tensor
+  {
+    using namespace functional;
+    Reduce(abs(_1), max(_1, _2), 0.0f, tempVar_, tflat);
+    S = tempVar_->get(0);
+  }
+
+  // optimze the scaling factor S
+  for(int i = 0; i < optSteps_; i++) {
```

Comment:
A practical question here: This shares the scaling across all weights, right? But are the weights not possibly in different numeric scales? Or will the shared scaling factor kind of impose that the model moves towards a regime where they are in similar range?

Reply:
each weight matric has its own scaling factor, to accommodate the issue of them having different ranges.

Reply:
I assume they are in tempVar_. Do I see correctly that this is not stateful? I.e. we don't need to store/re-store that for continued training?